include/oneapi/internal/execution_context_sycl.h

```diff
@@ -41,164 +42,105 @@ namespace interface1
-
-        ~KernelCacheEntry() {}
 
-        void setKernel(KernelPtr kernel, const char *name)
+        _currentProgramRef = programHashTable.get(key, local_status).get();
+        if (!local_status.ok())
         {
-            _name = name;
-            _kernel = kernel;
+            services::internal::tryAssignStatus(status, local_status);
+            return;
         }
 
-        KernelPtr getKernel()
-        {
-            return _kernel;
-        }
+        _executionTarget = target;
+    }
 
-        const char* getName()
+    KernelPtr getKernel(const char * kernelName, services::Status * status = nullptr) DAAL_C11_OVERRIDE
+    {
+        if (_currentProgramRef == nullptr)
         {
-            return _name.c_str();
+            services::internal::tryAssignStatus(status, services::Status().add(services::ErrorExecutionContextCache));
+            return KernelPtr();
         }
-    };
-
-public:
-    explicit OpenClKernelFactory(cl::sycl::queue & deviceQueue) : _clProgramRef(nullptr), _executionTarget(ExecutionTargetIds::unspecified), _deviceQueue(deviceQueue)
-    {}
 
-    void build(ExecutionTargetId target, const char * key, const char * program, const char * options = "",
-               services::Status * status = nullptr) DAAL_C11_OVERRIDE
-    {
-        // TODO: Thread safe?
-        // TODO Rework of "cache"
+        services::Status local_status;
 
-        uint64_t id = hash(key) % SIZE_CACHE_PROGRAM;
+        services::String key = _currentProgramRef->getName();
+        key.add(kernelName);
 
-        while (_clProgramCache[id].getProgram() != nullptr && strcmp(_clProgramCache[id].getName(), key) != 0)
+        bool res = kernelHashTable.contain(key, local_status);
+        if (!local_status.ok())
         {
-            id = (id + 1) % SIZE_CACHE_PROGRAM;
+            services::internal::tryAssignStatus(status, local_status);
+            return KernelPtr();
         }
-
-        if (_clProgramCache[id].getProgram())
+        if (res)
         {
-            _clProgramRef    = _clProgramCache[id].getProgram();
-            _executionTarget = target;
+            auto kernel = kernelHashTable.get(key, local_status);
+            services::internal::tryAssignStatus(status, local_status);
+            return kernel;
         }
-        else
-        {
-            _clProgramCache[id].setProgram(new OpenClProgramRef(_deviceQueue.get_context().get(),
-                                           _deviceQueue.get_device().get(), key, program, options, status));
-            if (status != nullptr && !status->ok())
-            {
-                return;
-            }
-            _clProgramRef    = _clProgramCache[id].getProgram();
-            _executionTarget = target;
-        }
-    }
-
-    KernelPtr getKernel(const char * kernelName, services::Status * status = nullptr) DAAL_C11_OVERRIDE
-    {
-        DAAL_ASSERT(_clProgramRef);
-        DAAL_ASSERT(*_clProgramRef);
-        KernelPtr kernelPtr;
 
```

Comment:
Too many blank lines; Also, I would prefer to wrap code on lines 118-134 into `else` clause to make the logic more clear and explicit.